core/src/main/java/io/grpc/internal/KeepAliveManager.java

```diff
@@ -131,27 +133,34 @@ public void run() {
    * Creates a KeepAliverManager.
    */
   public KeepAliveManager(ManagedClientTransport transport, ScheduledExecutorService scheduler,
-                          long keepAliveDelayInNanos, long keepAliveTimeoutInNanos) {
-    this.transport = Preconditions.checkNotNull(transport, "transport");
-    this.scheduler = Preconditions.checkNotNull(scheduler, "scheduler");
-    this.ticker = SYSTEM_TICKER;
-    // Set a minimum cap on keepalive dealy.
-    this.keepAliveDelayInNanos = Math.max(MIN_KEEPALIVE_DELAY_NANOS, keepAliveDelayInNanos);
-    this.keepAliveTimeoutInNanos = keepAliveTimeoutInNanos;
-    nextKeepaliveTime = ticker.read() + keepAliveDelayInNanos;
+                          long keepAliveDelayInNanos, long keepAliveTimeoutInNanos,
+                          boolean keepAliveDuringTransportIdle) {
+    this(transport, scheduler, SYSTEM_TICKER,
+        // Set a minimum cap on keepalive dealy.
+        Math.max(MIN_KEEPALIVE_DELAY_NANOS, keepAliveDelayInNanos), keepAliveTimeoutInNanos,
+        keepAliveDuringTransportIdle);
   }
 
   @VisibleForTesting
   KeepAliveManager(ManagedClientTransport transport, ScheduledExecutorService scheduler,
-                   Ticker ticker, long keepAliveDelayInNanos, long keepAliveTimeoutInNanos) {
+                   Ticker ticker, long keepAliveDelayInNanos, long keepAliveTimeoutInNanos,
+                   boolean keepAliveDuringTransportIdle) {
     this.transport = Preconditions.checkNotNull(transport, "transport");
     this.scheduler = Preconditions.checkNotNull(scheduler, "scheduler");
     this.ticker = Preconditions.checkNotNull(ticker, "ticker");
     this.keepAliveDelayInNanos = keepAliveDelayInNanos;
     this.keepAliveTimeoutInNanos = keepAliveTimeoutInNanos;
+    this.keepAliveDuringTransportIdle = keepAliveDuringTransportIdle;
     nextKeepaliveTime = ticker.read() + keepAliveDelayInNanos;
   }
 
+  /** Start keepalive monitoring. */
+  public synchronized void start() {
```

Comment:
I don't think it is necessary. On server side, `channelActive` is always called at the beginning, we can put an `onTransportActive` there.